Tool (#1)
jayfish0/llama_index: llama-index-integrations/tools/llama-index-tools-agentql/README.md

# AgentQL Tool

[AgentQL](https://www.agentql.com/) is a tool for web agents to interact with web elements and extracting data from web pages using Natural language or [AgentQL query](https://docs.agentql.com/agentql-query) for more precise action.

> **Warning**

> > Only support async functions and playwright browser APIs.

Author: This PR explains reason for async only: https://github.com/run-llama/llama_index/pull/17808

Reviewer: Pls, add this context in the doc itself, people will have the same question I had

Author: I added the link to the warning, it would be pretty long in the docs to explain everything

## Installation

```bash
pip install llama-index-tools-agentql
```

And you should configure credentials by setting the following environment variables:

> - AGENTQL_API_KEY

Reviewer 2: are we able to link to `dev.agentql.com` here in docs?

## Overview

> AgentQL provide the following three tools:

Reviewer 2: provide -> provides

- **`extract_web_data`**: Extract structured data in JSON form from webpage provided by a url with either natural language description or AgentQL query.

> - **`extract_web_data_with_browser`**: Extracted structured data in JSON from the current browser webpage with either natural language description or AgentQL query. (Must use with a browser instance)

Reviewer 2: Extracted -> Extract

naming convention question.  if we used with_browser elsewhere already, not a big deal, but this and below to me sounds more like "from_browser" if it requires a local browser instance

Author: oh yea for langchain we have extractDataBrowserTool, but in this case its a function call. I do agree extract_web_data_from_browser make more sense. Will change to this

- **`extract_web_element_with_browser`**: Extract the CSS selector of a web element from the current browser webpage with natural language description. (Must use with a browser instance)

## Setup

> In order to use this tool, you need to have a async Playwright browser instance. You can hook one up using the `create_async_playwright_browser` method:

Reviewer 2: a async -> an async

# AgentQL Tool

> [AgentQL](https://www.agentql.com/) is a tool for web agents to interact with web elements and extracting data from web pages using Natural language or [AgentQL query](https://docs.agentql.com/agentql-query) for more precise action.

Reviewer 3: ```suggestion
[AgentQL](https://agentql.com/) is a tool for web agents to interact with web elements and extract structured data from web pages using Natural Language or a precise [AgentQL query](https://docs.agentql.com/concepts/query-language).
```

> **Warning**

> > Only support async functions and playwright browser APIs.

Reviewer 3: ```suggestion
> AgentQL only supports async functions and Playwright's browser APIs.
```

## Installation

```bash
pip install llama-index-tools-agentql
```

And you should configure credentials by setting the following environment variables:

> - AGENTQL_API_KEY

Reviewer 3: ```suggestion

First, install Llama Index's AgentQL tools by running the following in your terminal:

```bash
pip install llama-index-tools-agentql
```

Next configure credentials. Set `AGENTQL_API_KEY` to an API key from our [dev portal](https://dev.agentql.com)
```

> You can acquire an API key from our [dev portal](https://dev.agentql.com).

Reviewer 3: ```suggestion
```

## Overview

AgentQL provides the following three tools:

- **`extract_web_data`**: Extract structured data in JSON form from webpage provided by a url with either natural language description or AgentQL query.

- **`extract_web_data_from_browser`**: Extracted structured data in JSON from the current browser webpage with either natural language description or AgentQL query. (Must use with a browser instance)

> - **`extract_web_element_from_browser`**: Extract the CSS selector of a web element from the current browser webpage with natural language description. (Must use with a browser instance)

Reviewer 3: Shouldn't this be `get_web_element_from_browser`?

Author: @colriot double checking, I'll rename extract to get for web element in langchain as well

> - **`extract_web_element_from_browser`**: Extract the CSS selector of a web element from the current browser webpage with natural language description. (Must use with a browser instance)

Reviewer 3: ```suggestion
- **`extract_web_data`** extracts structured data as JSON from a web page using either natural language description or an [AgentQL query](https://docs.agentql.com/agentql-query/query-intro).

- **`extract_web_data_from_browser`** extracts structured data as JSON from the web page currently open in the browser using either natural language description or an [AgentQL query](https://docs.agentql.com/agentql-query/query-intro). (Requires a browser instance)

- **`extract_web_element_from_browser`** extracts the CSS selector of a web element from the web page currently open in the browser using a natural language description. (Requires a browser instance.)
```

## Setup

In order to use this tool, you need to have an async Playwright browser instance. You can hook one up using the `create_async_playwright_browser` method:

```python
browser = await AgentQLToolSpec.create_async_playwright_browser(headless=False)
agentql_tool = AgentQLToolSpec.from_async_browser(browser)
```

## Usage

### Extract data from a webpage

```python
await agentql_tool.extract_web_data(
    "https://www.agentql.com/blog",
    query="{ blogs[] { title url author date }}",
)
```

### Extract data from the current browser instance

> **Note**
> Agentql browser tools are best used along with [playwright tools](https://llamahub.ai/l/tools/llama-index-tools-playwright?from=).

```python
await playwright_tool.navigate_to("https://www.agentql.com/blog")
await agentql_tool.extract_web_data_from_browser(
    prompt="Extract all the blog titles and urls from the current page.",
)
```

> ### Extract the CSS selector of a web element from the current browser instance

Reviewer 3: Get?

# llama-index-tools-agentql

[AgentQL](https://www.agentql.com/) provides web interaction and structured data extraction from any web page using an [AgentQL query](https://docs.agentql.com/agentql-query) or a Natural Language prompt. AgentQL can be used across multiple languages and web pages without breaking over time and change.

> **Warning**
> Only support async functions and playwright browser APIs.

## Installation

```bash
pip install llama-index-tools-agentql
```

You also need to configure the `AGENTQL_API_KEY` environment variable. You can acquire an API key from our [Dev Portal](https://dev.agentql.com).

## Overview

AgentQL provides the following three tools:

- **`extract_web_data`**: Extracts structured data as JSON from a web page given a URL using either an [AgentQL query](https://docs.agentql.com/agentql-query/query-intro) or a Natural Language description of the data.

- **`extract_web_data_from_browser`**: Extracts structured data as JSON from the active web page in a browser using either an [AgentQL query](https://docs.agentql.com/agentql-query/query-intro) or a Natural Language description. **This tool must be used with a Playwright browser.**

> - **`get_web_element_from_browser`**: Finds a web element on the active web page in a browser using a Natural Language description and returns its CSS selector for further interaction. **This tool must be used with a Playwright browser.**

Reviewer 3: Align these to https://docs.google.com/document/d/1U6Xa0a7cs70rVFB5M4edmF252NyUzJcFIZDVz1zgAbs/edit?tab=t.76ce0iz2zkkr

You can learn more about how to use AgentQL tools in this [Jupyter notebook](https://github.com/run-llama/llama_index/blob/main/llama-index-integrations/tools/llama-index-tools-agentql/examples/AgentQL_browser_agent.ipynb).

### Extract data using REST API

```python
from llama_index.tools.agentql.base import AgentQLToolSpec

> agentql_tool = AgentQLToolSpec()

Reviewer: and where will be a browser provided from?

Author: In this case we are only using `extract_web_data_with_rest_api`, and therefore a browser is not needed. You are right I missed a toolspec initialization with the browser tools. Just added

await agentql_tool.extract_web_data(
    "https://www.agentql.com/blog",
    query="{ blogs[] { title url author date }}",
)
```

### Work with data and web elements using browser

#### Setup

In order to use the `extract_web_data_from_browser` and `get_web_element_from_browser`, you need to have a Playwright browser instance. If you do not have an active instance, you can initiate one using the `create_async_playwright_browser` utility method from LlamaIndex's Playwright ToolSpec.

> **Note**

> > Agentql browser tools are best used along with LlamaIndex's [Playwright](https://docs.llamaindex.ai/en/stable/api_reference/tools/playwright/) tools.

Reviewer: ```suggestion
> Agentql browser tools are best used along with LlamaIndex's [Playwright tools](https://docs.llamaindex.ai/en/stable/api_reference/tools/playwright/).
```

```python
from llama_index.tools.playwright.base import PlaywrightToolSpec

async_browser = await PlaywrightToolSpec.create_async_playwright_browser(

>     headless=False

Reviewer: this is default, can be removed?

)
```

You can also use an existing browser instance via Chrome DevTools Protocol (CDP) connection URL:

```python
p = await async_playwright().start()
async_browser = await p.chromium.connect_over_cdp("CDP_CONNECTION_URL")
```

#### Extract data from the active browser page

```python
playwright_tool = PlaywrightToolSpec(async_browser=async_browser)

await playwright_tool.navigate_to("https://www.agentql.com/blog")
await agentql_tool.extract_web_data_from_browser(

>     prompt="Extract all the blog titles and urls from the current page.",

Reviewer: It's not a task, it's a description of data. Same comment as in langchain)
```suggestion
    prompt="The blog posts with title, url, date of post and author",
```

)
```

#### Find a web element on the active browser page

```python

> next_page_button = await agentql_tool.extract_web_element_from_browser(

Reviewer: ?
```suggestion
next_page_button = await agentql_tool.get_web_element_from_browser(
```

# llama-index-tools-agentql

[AgentQL](https://www.agentql.com/) provides web interaction and structured data extraction from any web page using an [AgentQL query](https://docs.agentql.com/agentql-query) or a Natural Language prompt. AgentQL can be used across multiple languages and web pages without breaking over time and change.

> **Warning**

> > Only support async functions and playwright browser APIs, please refer to the following PR for more details: https://github.com/run-llama/llama_index/pull/17808

Reviewer 3: ```suggestion
> Only supports async functions and playwright browser APIs, please refer to the following PR for more details: https://github.com/run-llama/llama_index/pull/17808
```

> **Note**
> Agentql browser tools are best used along with LlamaIndex's [Playwright tools](https://docs.llamaindex.ai/en/stable/api_reference/tools/playwright/).

```python
from llama_index.tools.playwright.base import PlaywrightToolSpec

async_browser = await PlaywrightToolSpec.create_async_playwright_browser()
```

You can also use an existing browser instance via Chrome DevTools Protocol (CDP) connection URL:

```python
p = await async_playwright().start()
async_browser = await p.chromium.connect_over_cdp("CDP_CONNECTION_URL")
```

#### Extract data from the active browser page

```python
from llama_index.tools.agentql import AgentQLBrowserToolSpec

playwright_tool = PlaywrightToolSpec(async_browser=async_browser)
await playwright_tool.navigate_to("https://www.agentql.com/blog")

agentql_browser_tool = AgentQLBrowserToolSpec(async_browser=async_browser)
await agentql_browser_tool.extract_web_data_from_browser(

>     prompt="the blog posts' title and url",

Reviewer: ```suggestion
    prompt="the blog posts with title and url",
```

- **`get_web_element_from_browser`**: Finds a web element on the active web page in a browser using a Natural Language description and returns its CSS selector for further interaction. **This tool must be used with a Playwright browser.**

You can learn more about how to use AgentQL tools in this [Jupyter notebook](https://github.com/run-llama/llama_index/blob/main/llama-index-integrations/tools/llama-index-tools-agentql/examples/AgentQL_browser_agent.ipynb).

### Extract data using REST API

```python
from llama_index.tools.agentql import AgentQLRestAPIToolSpec

agentql_rest_api_tool = AgentQLRestAPIToolSpec()
await agentql_rest_api_tool.extract_web_data_with_rest_api(
    url="https://www.agentql.com/blog",
    query="{ posts[] { title url author date }}",
)
```

### Work with data and web elements using browser

#### Setup

In order to use the `extract_web_data_from_browser` and `get_web_element_from_browser`, you need to have a Playwright browser instance. If you do not have an active instance, you can initiate one using the `create_async_playwright_browser` utility method from LlamaIndex's Playwright ToolSpec.

> **Note**

> > Agentql browser tools are best used along with LlamaIndex's [Playwright tools](https://docs.llamaindex.ai/en/stable/api_reference/tools/playwright/).

Reviewer: AgentQL
